README.md
# gggibbous

Moon charts, a pie chart alternative for two groups in ggplot2

![gggibbous](man/figures/gggibbous.png)

## Installation

`gggibbous` can be installed from CRAN:

``` r
install.packages("gggibbous")
```

The development version can be installed from GitHub with the `devtools`
package:

``` r
devtools::install_github("mnbram/gggibbous")
```

## Moon charts and pie charts

A pie chart divides a circle into multiple sections where the arc
lengths (and so also the areas) of the sections represent proportions of
a whole. A moon chart, similarly, divides a circle into sections where
the areas represent proportions of a whole, but in a moon chart the
areas are drawn as crescent or gibbous portions of a circle—like the
phases of the moon.

![](man/figures/README-unnamed-chunk-5-1.png)<!-- -->

The motivation behind using a moon chart instead of a pie chart is
primarily one of aesthetic choice. Note also that because the sections
of a moon chart are swept from one or the other side of the circle, they
are generally only appropriate for depicting one or two groups.

Moon charts are similar to Kosara’s (2019)\[1\] “circular slice” chart.
In studying subjects’ perception of percentages in different chart
types, the “circular slice” performed similarly to pie charts. Moon
charts differ from the “circular slice” in that the latter slides a
second disc of the same size over a base circle, more like a lunar
eclipse than the sphases of the moon. Both depend on area as the
perceptual cue, however.

## `gggibbous` and its usage

`gggibbous` extends the `ggplot2` data visualization package to provide
support for moon charts in R. Unlike the pie charts supported natively
by `coord_polar()` in R, moon charts in `gggibbous` do not require any
special coordinate system. They are drawn most similarly to points in
`ggplot2`: their position is defined by an `x` and a `y` coordinate and
their size is defined independently of the coordinate system, so they
always remain circular.

``` r
library(gggibbous)
#> Loading required package: ggplot2
```

``` r
ggplot(data.frame(x = 1:5, y = 1, size = 2^(0:4)), aes(x, y, size = size)) +
  geom_moon() +
  geom_point(y = 2) +
  lims(x = c(0.5, 5.5), y = c(0.5, 2.5)) +
  scale_size(range = c(5, 10))
```

![](man/figures/README-unnamed-chunk-6-1.png)<!-- -->

Two new aesthetics are also important in `geom_moon`: `ratio` and
`right`.

### The `ratio` aesthetic

`ratio` controls the proportion of the moon to be drawn. It must be
between 0 (a “new moon” where nothing is actually drawn) and 1 (a “full
moon”, i.e. a circle).

``` r
ggplot(data.frame(x = 1:5, y = 0, ratio = 0:4 * 0.25), aes(x = x, y = y)) +
  geom_moon(aes(ratio = ratio), size = 20, fill = "black") +
  geom_text(aes(y = y + 1, label = ratio)) +
  lims(x = c(0.5, 5.5), y = c(-1, 1.4)) +
  theme_void()
```

![](man/figures/README-unnamed-chunk-7-1.png)<!-- -->

### The `right` aesthetic

`right` takes a boolean value that controls whether the moon is “waxing”
or “waning”—that is, whether it is “filled” from the right or the left.

One way to make a “complete” moon with two colors is to use `right =
TRUE` for one color and `right = FALSE` for the other, with
complementary ratios. (See the examples below for some other
approaches.)

``` r
tidymoons <- data.frame(
  x = rep(1:3, 6),
  y = rep(rep(3:1, each = 3), 2),
  ratio = c(1:9 / 10, 9:1 / 10),
  right = rep(c(TRUE, FALSE), each = 9)
)

ggplot(tidymoons) +
  geom_moon(aes(x, y, ratio = ratio, right = right, fill = right)) +
  lims(x = c(0.5, 3.5), y = c(0.5, 3.5))
```

![](man/figures/README-unnamed-chunk-8-1.png)<!-- -->

## Legend key glyphs

`gggibbous` includes three key glyphs for different types of legends:
`draw_key_moon`, `draw_key_moon_left`, and `draw_key_full_moon`

**`draw_key_moon`**—the default in `geom_moon`—draws a gibbous moon with
`right = TRUE` (see above).

**`draw_key_moon_left`** draws a crescent moon from the left that is
complementary to the gibbous moon in `draw_key_moon`, which is useful
for combined legends:

``` r
ggplot(tidymoons, aes(x, y, ratio = ratio, right = right, size = 2^x)) +
  geom_moon(data = subset(tidymoons, right), fill = "violetred") +
  geom_moon(
    data = subset(tidymoons, !right), fill = "turquoise3",
    key_glyph = draw_key_moon_left
  ) +
  lims(x = c(0.5, 3.5), y = c(0.5, 3.5)) +
  scale_size("size", range = c(5, 10), breaks = 2^(1:3))
```

![](man/figures/README-unnamed-chunk-9-1.png)<!-- -->

**`draw_key_full_moon`** draws a circle. It is similar to the “point”
key glyph, but the size is calculated slightly differently, so it is
more appropriate if you want the size of legend moons and the size of
the plot moons to match.

``` r
ggplot(tidymoons) +
  geom_moon(
    aes(x, y, ratio = ratio, right = right, fill = right, size = 2^x),
    key_glyph = draw_key_full_moon
  ) +
  lims(x = c(0.5, 3.5), y = c(0.5, 3.5)) +
  scale_size("size", range = c(5, 10), breaks = 2^(1:3)) +
  scale_fill_manual(values = c("firebrick1", "dodgerblue2")) +
  theme(legend.box = "horizontal")
```

![](man/figures/README-unnamed-chunk-10-1.png)<!-- -->

## Worked examples

### Moon charts on maps

One common use of multiple pie charts is to represent proportions at
different coordinates on a map. The *x* and *y* dimensions are already
committed to the map coordinates, so proportional visualizations like
bar charts are more difficult. This is a perfect opportunity to try out
moon charts\!

Pie chart maps are popular in population genetics, so let’s look at an
example from that field. The `dmeladh` data\[2\] contains frequencies of
two variants of the *Adh* gene in Australian and Papua New Guinean fruit
fly populations. Many of these populations are close together, so we
have to deal with overplotting, which we do manually below.

``` r
dmeladh_adj <- dmeladh
dmeladh_adj$long <- dmeladh$Longitude + c(
  -2, 0, -2, 2, -3, 3, 3, 2, 3, 4, -2.5, -2.5, -1, -2, -2.5, -4, 2.5,
  5, 6, 7, 2, -7, -5.5, -3, 0, -7, -2, 3, 5.5, 0.5, -1, -1.5, -3, 2)
dmeladh_adj$lat <- dmeladh$Latitude + c(
  -2, 2, 0, 1, 0, 0, 0, 2, 0.5, -1, 1, -1.5, 2, 4, 1.5, 0, 2,
  1, -1, -3, -2, 1, -1, -2, -3, -2, -4, -3, -1, 1.5, 2, 2, -2, 0)

moonmap <- ggplot(dmeladh_adj, aes(long, lat)) +
  geom_polygon(
    data = map_data(
      "world", region = "(Australia)|(Indonesia)|(Papua New Guinea)"),
    aes(group = group),
    fill = "gray80"
  ) +
  geom_segment(aes(xend = Longitude, yend = Latitude), color = "gray20") +
  geom_point(aes(Longitude, Latitude), size = 0.75, color = "gray20") +
  scale_size(range = c(4, 10)) +
  coord_map(xlim = c(110, 160), ylim = c(-45, -5)) +
  theme_void() +
  theme(
    legend.position = c(0.05, 0.05),
    legend.direction = "horizontal",
    legend.justification = c(0, 0)
  )

moonmap +
  geom_moon(
    aes(ratio = AdhS / 100, size = N),
    right = FALSE, fill = "gold", color = "gold",
    key_glyph = draw_key_moon_left
  ) +
  geom_moon(
    aes(ratio = AdhF / 100, size = N),
    fill = "forestgreen", color = "forestgreen"
  )
```

![](man/figures/README-unnamed-chunk-11-1.png)<!-- -->

If we want to label the alleles in the legend explicitly, then we need
to map them to a group, which requires that we rearrange the data into a
“longer” (“tidy”) format.

``` r
tidyadh <- reshape(
  dmeladh_adj,
  varying = c("AdhF", "AdhS"),
  v.names = "percent",
  timevar = "allele",
  times = c("AdhF", "AdhS"),
  idvar = c("Locality", "Latitude", "Longitude", "long", "lat", "N"),
  direction = "long"
)
tidyadh$right <- rep(c(TRUE, FALSE), each = nrow(dmeladh_adj))

moonmap +
  geom_moon(
    data = tidyadh, key_glyph = draw_key_full_moon,
    aes(ratio = percent / 100, fill = allele, color = allele, right = right,
        size = N)
  ) +
  scale_fill_manual(values = c("forestgreen", "gold")) +
  scale_color_manual(values = c("forestgreen", "gold"))
```

![](man/figures/README-unnamed-chunk-12-1.png)<!-- -->

### Lunar data

Sometimes you just want to plot literal representations of the moon. The
`lunardist` data, adapted from NASA, contains the distance from the
Earth to the Moon for each day in 2019, as well as the dates (in UTC) of
each occurrence of the four principal phases of the moon. We can plot
those principal phases using moon charts (which in this case are
identical to pie charts).

``` r
moonphase <- subset(lunardist, !is.na(phase))
moonphase$percent <- ifelse(
  moonphase$phase == "new", 0, ifelse(moonphase$phase == "full", 1, 0.5))

ggplot(lunardist, aes(date, distance)) +
  geom_line() +
  # Plotting the lower layer as a full circle also works in most cases
  geom_moon(data = moonphase, ratio = 1, size = 5, fill = "black") +
  geom_moon(
    data = moonphase, aes(ratio = percent),
    size = 5, fill = "yellow", right = moonphase$phase == "first quarter"
  )
```

![](man/figures/README-unnamed-chunk-13-1.png)<!-- -->

### Harvey balls

“Harvey balls” are essentially pie charts used for qualitative
comparisons, often in tabular format. We can use moon charts for the
same purpose.

First, let’s make up some data:

``` r
rest_names <- c(
  "Anscombe's Luncheonette", "Chai Squared", "Tukey's Honest Southern Diner",
  "Bagels ANOVA", "Spearmint Row"
)
restaurants <- data.frame(
  Restaurant = factor(rest_names, levels = rest_names),
  Food = c(5, 3, 4, 4, 1),
  Decor = c(2, 5, 3, 1, 5),
  Service = c(4, 2, 3, 3, 5),
  Price = c(4, 5, 2, 5, 2)
)
```

As a regular table:

| Restaurant                    | Food | Decor | Service | Price |
| :---------------------------- | :--: | :---: | :-----: | :---: |
| Anscombe’s Luncheonette       |  5   |   2   |    4    |   4   |
| Chai Squared                  |  3   |   5   |    2    |   5   |
| Tukey’s Honest Southern Diner |  4   |   3   |    3    |   2   |
| Bagels ANOVA                  |  4   |   1   |    3    |   5   |
| Spearmint Row                 |  1   |   5   |    5    |   2   |

And now as a table with Harvey moons:

``` r
# First we reshape the data into "long" format to facilitate plotting
rest_cats <- c("Food", "Decor", "Service", "Price")
tidyrest <- reshape(
  restaurants,
  varying = rest_cats,
  v.names = "Score",
  timevar = "Category",
  times = factor(rest_cats, levels = rest_cats),
  idvar = "Restaurant",
  direction = "long"
)

ggplot(tidyrest, aes(0, 0)) +
  geom_moon(aes(ratio = (Score - 1) / 4), fill = "black") +
  geom_moon(aes(ratio = 1 - (Score - 1) / 4), right = FALSE) +
  facet_grid(Restaurant ~ Category, switch = "y") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    strip.text.y.left = element_text(angle = 0, hjust = 1),
    axis.text = element_blank(),
    axis.title = element_blank()
  )
```

![](man/figures/README-unnamed-chunk-16-1.png)<!-- -->

## References

1.   Kosara, R. 2019. Circular Part-to-Whole Charts Using the Area
    Visual Cue. EuroVis 2019 - Short Papers.
    <https://doi.org/10.2312/evs.20191163>

2.  Oakeshott, J.G., et al. 1982. Alcohol dehydrogenase and
    glycerol-3-phosphate dehydrogenase clines in *Drosophila
    melanogaster* on different continents. Evolution, 36(1): 86-96.


NEWS.md
# gggibbous 0.1.1

* Updated vignette and README to cite Robert Kosara's related work

* vdiffr is now called conditionally in tests so as to avoid issues on platforms
  that do not have X11

# gggibbous 0.1.0

* Initial release

tests/testthat.R
library(testthat)
library(gggibbous)

test_check("gggibbous")


tests/testthat/test-vdiffr.R
context("test-vdiffr")

# Thanks to Lionel Henry for the suggestion on using vdiffr conditionally
# See https://github.com/lionel-/ggstance for the original
expect_doppelganger <- function(title, fig, path = NULL, ...) {
  testthat::skip_if_not_installed("vdiffr")
  vdiffr::expect_doppelganger(title, fig, path = path, ...)
}

five_moons <- data.frame(
  x = 1:5,
  y = 1,
  size = 2^(1:5),
  ratio = 0:4 * 0.25,
  group = letters[1:5]
)

moon_size <- ggplot(five_moons, aes(x = x, y = y, size = size)) +
  geom_moon(ratio = 0.5, right = FALSE) +
  scale_size_area(max_size = 30)

expect_doppelganger("moons scale correctly", moon_size)

moon_missing <- ggplot(
  five_moons, aes(x = x, y = y, ratio = ratio, fill = group, color = group)) +
  geom_moon(size = 10)

expect_doppelganger("zero ratio moons are invisible", moon_missing)

full_moons <- ggplot(five_moons, aes(x = x, y = y)) +
  geom_moon(aes(ratio = ratio), fill = "black", size = 10) +
  geom_moon(aes(ratio = 1 - ratio), right = FALSE, size = 10)

expect_doppelganger("moons can be complemented", full_moons)


tests/testthat/test-moonGrob.R
context("test-moonGrob")

test_that("moon grobs are grobs", {
  expect_is(moonGrob(x = 0, y = 0), "grob")
})

test_that("ratio must be between 0 and 1", {
  expect_error(grid.moon(x = 0, y = 0, ratio = 2))
  expect_error(grid.moon(x = 0, y = 0, ratio = -1))
})


R/data.R
#' \emph{Adh} allele frequencies in Australasian \emph{Drosophila melanogaster}
#' 
#' This data set contains allele frequencies for the "fast" and "slow" variants
#' of the enzyme alcohol dehydrogenase in Australasian (mostly Australian)
#' populations of \emph{Drosophila melanogaster}. The data are taken from
#' Oakeshott, J.G., et al. 1982. Alcohol dehydrogenase and glycerol-3-phosphate
#' dehydrogenase clines in \emph{Drosophila melanogaster} on different
#' continents. Evolution, 36(1): 86-96.
#' 
#' @format A data frame with 34 rows and 6 variables:
#' \describe{
#'   \item{Locality}{location of population sample}
#'   \item{Latitude}{latitude of population sample}
#'   \item{Longitude}{longitude of population sample}
#'   \item{N}{number of samples}
#'   \item{AdhF}{percent of samples with \emph{Adh^F} allele, as an integer}
#'   \item{AdhS}{percent of samples with \emph{Adh^S} allele, as an integer}
#' }
"dmeladh"

#' Lunar distances and principal phases for 2019
#'
#' This data set contains the distance from the Earth to the Moon for each day
#' in 2019, as well as the dates (in UTC) of each occurrence of the four
#' principal phases of the moon. The data are adapted from NASA.
#' 
#' @format A data frame with 365 rows and 3 variables:
#' \describe{
#'   \item{date}{Date}
#'   \item{distance}{Distance from the Earth to the Moon in kilometers}
#'   \item{phase}{Principal phase of the moon (full, new, first quarter, third
#'   quarter) or NA if no principal phase on that date}
#' }
#' 
#' @source \url{https://svs.gsfc.nasa.gov/4442}
"lunardist"

R/moonGrob.R
#' Draw a moon
#' 
#' Functions to create and draw crescent or gibbous moon shapes.
#' 
#' Both functions create a moon grob (a graphical object describing a crescent
#' or gibbous moon), but only \code{grid.moon} draws the moon (and then only
#' if draw is TRUE).
#' 
#' These functions calculate a set of points describing the perimeters of the
#' moons and pass these points on to \code{grid::polygonGrob}.
#' 
#' The units in \code{default.units} and \code{size.units} can be different;
#' \code{grid} will add them together appropriately before drawing.
#' 
#' @param ratio A numeric vector with values from 0 to 1 specifying the
#'   proportion of the moons to draw.
#' @param right A boolean vector specifying whether the moon should be filled
#'   from the right (TRUE) or left (FALSE).
#' @param r A numeric vector specifying the radii of the circles describing
#'   the moons.
#' @param angle Not used.
#' @param size.units A string indicating the units to use for the radii of the
#'   moons.
#' @param ... Arguments passed on to \code{grid::polygonGrob}.
#' @inheritParams grid::polygonGrob
#' @return A grob object.
#' @export
#' @examples
#' grid::grid.newpage()
#' grid.moon(x = 1:3 * 0.25, y = rep(0.5, 3), ratio = 1:3 * 0.25, r = 10)
moonGrob <- function(
  x, y, ratio = 0.25, right = TRUE, r = 10, angle = 0,
  default.units = "npc", size.units = "mm", ...
) {
  
  stopifnot(is.numeric(x))
  stopifnot(is.numeric(y))
  stopifnot(is.numeric(ratio))
  if (any(ratio < 0) | any(ratio > 1)) {
    stop("Error: ratios must be between 0 and 1")
  }
  if (!is.logical(right)) {
    stop("Error: `right` must be a logical vector")
  }
  stopifnot(is.logical(right))
  stopifnot(is.numeric(r))
  # stopifnot(is.numeric(angle))
  stopifnot(is.character(default.units))
  stopifnot(is.character(default.units))
  stopifnot(is.character(size.units))
  stopifnot(default.units %in% .grid.unit.list)
  stopifnot(size.units %in% .grid.unit.list)
  stopifnot(length(x) == length(y))
  stopifnot(length(ratio) %in% c(1, length(x)))
  stopifnot(length(right) %in% c(1, length(x)))
  stopifnot(length(r) %in% c(1, length(x)))
  stopifnot(length(angle) %in% c(1, length(x)))
  
  if (length(ratio) == 1) {
    ratio <- rep(ratio, length(x))
  }
  if (length(right) == 1) {
    right <- rep(right, length(x))
  }
  if (length(r) == 1) {
    r <- rep(r, length(x))
  }
  if (length(angle) == 1) {
    angle <- rep(angle, length(x))
  }
  
  coords_list <- mapply(
    moon_coords,
    x = x, y = y, ratio = ratio, right = right, r = r, angle = angle,
    MoreArgs = list(
      default.units = default.units, size.units = size.units
    )
  )
  
  x_coords <- do.call(grid::unit.c, coords_list["x_coords",])
  y_coords <- do.call(grid::unit.c, coords_list["y_coords",])
  
  grid::polygonGrob(
    x_coords, y_coords,
    id.lengths = sapply(coords_list["x_coords",], length),
    default.units = default.units, ...
  )
  
}


#' @rdname moonGrob
#' @export
grid.moon <- function(..., draw = TRUE) {
  mg <- moonGrob(...)
  if (draw) {
    grid::grid.draw(mg)
  }
  invisible(mg)
}


moon_coords <- function(
  x, y, ratio, right, r, angle, default.units, size.units
) {
  if (!grid::is.unit(x)) {
    x <- grid::unit(x, default.units)
  }
  if (!grid::is.unit(y)) {
    y <- grid::unit(y, default.units)
  }
  
  # If the ratio is 0, then we don't want to draw a polygon at all
  if (ratio == 0) {
    return(list(x_coords = x, y_coords = y))
  }
  
  # The circles and ellipses in gggibbous are actually grid's xspline
  # approximation of a cubic B-spline approximation of a Bezier
  # approximation of an ellipse. As such, the "magic" number used here
  # to determine the control points is slightly different than what one would
  # use for a true Bezier curve. This magic number has been determined
  # empirically by eye; it could probably be further refined.
  magic <- 0.571 # Magic number for Bezier approximation of ellipse
  
  # Calculate points of Bezier approximations of circle and ellipse
  # that define the shape.
  # First center the circle and ellipse at the origin (using size.units),
  # then we will translate everything to the right location (using
  # default.units).
  bez_circ_top <- grid::bezierPoints(grid::bezierGrob(
    c(0, magic * r, r, r),
    c(r, r, magic * r, 0),
    default.units = size.units
  ))
  bez_circ_bot <- grid::bezierPoints(grid::bezierGrob(
    c(r, r, magic * r, 0),
    c(0, -magic * r, -r, -r),
    default.units = size.units
  ))
  
  # Calculate scaling factor for ellipse
  e_scale <- -abs(2 * ratio - 1)
  # Crescent or gibbous?
  x_dir <- ifelse(ratio > 0.5, 1, -1)
  
  bez_elli_bot <- grid::bezierPoints(grid::bezierGrob(
    c(0, x_dir * magic * r * e_scale,
      x_dir * r * e_scale, x_dir * r * e_scale),
    c(-r, -r, -magic * r, 0),
    default.units = size.units
  ))
  bez_elli_top <- grid::bezierPoints(grid::bezierGrob(
    c(x_dir * r * e_scale, x_dir * r * e_scale,
      x_dir * magic * r * e_scale, 0),
    c(0, magic * r, r, r),
    default.units = size.units
  ))
  
  # Get x and y coordinates for whole perimeter
  poly_x <- grid::unit.c(
    bez_circ_top$x,
    bez_circ_bot$x[2:length(bez_circ_bot$x)],
    bez_elli_bot$x[2:length(bez_elli_bot$x)],
    bez_elli_top$x[2:(length(bez_elli_top$x) - 1)]
  )
  poly_y <- grid::unit.c(
    bez_circ_top$y,
    bez_circ_bot$y[2:length(bez_circ_bot$y)],
    bez_elli_bot$y[2:length(bez_elli_bot$y)],
    bez_elli_top$y[2:(length(bez_elli_top$y) - 1)]
  )
  
  # For a left-side moon, flip the x values around the origin
  if(right == FALSE) {
    poly_x <- grid::unit(0, size.units) - poly_x
  }
  
  # Translate the shape to the specified center point
  trans_x <- poly_x + x
  trans_y <- poly_y + y
  
  # TODO: Implement rotation
  
  list(x_coords = trans_x, y_coords = trans_y)
}


.grid.unit.list <- c(
  "npc", "cm", "inches", "lines", "native", "null", "snpc", "mm", "points",
  "picas", "bigpts", "dida", "cicero", "scaledpts", "strwidth", "strheight",
  "strascent", "strdescent", "vplayoutwidth", "vplayoutheight", "char", "grobx",
  "groby", "grobwidth", "grobheight", "grobascent", "grobdescent", "mylines",
  "mychar", "mystrwidth", "mystrheight"
)

R/gggibbous.R
#' gggibbous: Moon charts, a pie chart alternative for two groups
#'
#' Moon charts are like pie charts except that the proportions are shown as
#' crescent or gibbous portions of a circle, like the lit and unlit portions of
#' the moon. As such, they work best with only one or two groups.
#' \code{gggibbous} extends \code{ggplot2} to allow for plotting multiple moon
#' charts in a single panel and does not require a square coordinate system.
#' 
#' The workhorse function is \code{geom_moon}, which adds a moon chart layer to
#' a ggplot2 plot.
#' The \code{draw_key_moon}, \code{draw_key_moon_left}, and
#' \code{draw_key_full_moon} functions provides legend key glyphs for plots that
#' use \code{geom_moon}.
#' There are also functions for the raw \code{grid} grobs: \code{grid.moon} and
#' \code{moonGrob}.
#' 
#' For more information, see the \code{gggibbous} vignette.
#' 
#' @docType package
#' @name gggibbous
NULL

sticker/sticker.R
library(grid)
library(showtext)
library(svglite)

font_add_google("Economica", "economica")
showtext_auto()

hex_grid <- function(fontsize) {
  grid.newpage()
  hex_r <- 209 / 2
  hex_y <- hex_r * sqrt(3) / 2
  grid.polygon(
    x = unit(c(-hex_y, -hex_y, 0, hex_y, hex_y, 0),
             "bigpts") + unit(0.5, "npc"),
    y = unit(c(hex_r / 2, -hex_r / 2, -hex_r,
               -hex_r / 2, hex_r / 2, hex_r),
             "bigpts") + unit(0.5, "npc"),
    gp = gpar(fill = "black", lwd = 0.5)
  )
  moon_pct <- 0.26
  grid.moon(
    x = 0.5, y = 0.5, ratio = moon_pct, right = FALSE,
    r = 75, default.units = "npc", size.units = "bigpts",
    gp = gpar(fill = "lemonchiffon", color = "black", lwd = 1)
  )
  grid.moon(
    x = 0.5, y = 0.5, ratio = 1 - moon_pct,
    r = 75, default.units = "npc", size.units = "bigpts",
    gp = gpar(fill = "gray15", color = "gray15", lwd = 1)
  )
  grid.text(
    x = unit(0.5, "npc") + unit(-52.5, "bigpts"), y = unit(0.5, "npc"),
    label = "gg", default.units = "npc",
    gp = gpar(fontfamily = "economica", fontsize = fontsize, col = "black")
  )
  grid.text(
    x = unit(0.5, "npc") + unit(17, "bigpts"), y = unit(0.5, "npc"),
    label = "gibbous", default.units = "npc",
    gp = gpar(fontfamily = "economica", fontsize = fontsize, col = "white")
  )
}


png("sticker/gggibbous.png", width = 181, height = 209, bg = NA)
hex_grid(32)
dev.off()

# Font sizes don't make sense
# It seems that a difference in dpi rendering between graphical devices means
# that we need the SVG font size to be 4/3 (96/72) times bigger than for PNG
svglite("sticker/gggibbous.svg", width = 181/72, height = 209/72,
        bg = "transparent")
hex_grid(32 * 4 / 3)
dev.off()


R/geom_moon.R
#' Moon key glyph for legends
#'
#' Draws the legend key glyphs used in \code{geom_moon}.
#'
#' \code{draw_key_moon} (the default in \code{geom_moon}) draws a gibbous moon
#' filled from the right. \code{draw_key_moon_left} draws a crescent moon from
#' the right. \code{draw_key_full_moon} draws a circle, which is very similar to
#' \code{draw_key_point} in \code{ggplot2}, but the size is calculated slightly
#' differently and the default aesthetics differ.
#' 
#' @return A grid grob.
#' 
#' @inheritParams ggplot2::draw_key
#' @export
draw_key_moon <- function(data, params, size) {
  d_size <- ifelse(is.null(data$size), 10, data$size)
  d_col <- ifelse(is.null(data$colour), "black", data$colour)
  d_fill <- ifelse(is.null(data$fill), "white", data$fill)
  d_stroke <- ifelse(is.null(data$stroke), 0.25, data$stroke)
  d_ltype <- ifelse(is.null(data$linetype), "solid", data$linetype)
  moonGrob(
    0.5, 0.5, r = d_size / 2 * 0.75, ratio = 0.75,
    gp = grid::gpar(
      col = scales::alpha(d_col, data$alpha),
      fill = scales::alpha(d_fill, data$alpha),
      lwd = d_stroke * ggplot2::.stroke,
      lty = d_ltype
    )
  )
}

#' @rdname draw_key_moon
#' @export
draw_key_moon_left <- function(data, params, size) {
  d_size <- ifelse(is.null(data$size), 10, data$size)
  d_col <- ifelse(is.null(data$colour), "black", data$colour)
  d_fill <- ifelse(is.null(data$fill), "white", data$fill)
  d_stroke <- ifelse(is.null(data$stroke), 0.25, data$stroke)
  d_ltype <- ifelse(is.null(data$linetype), "solid", data$linetype)
  moonGrob(
    0.5, 0.5, r = d_size / 2 * 0.75, ratio = 0.25,
    gp = grid::gpar(
      col = scales::alpha(d_col, data$alpha),
      fill = scales::alpha(d_fill, data$alpha),
      lwd = d_stroke * ggplot2::.stroke,
      lty = d_ltype
    ),
    right = FALSE
  )
}

#' @rdname draw_key_moon
#' @export
draw_key_full_moon <- function(data, params, size) {
  d_size <- ifelse(is.null(data$size), 10, data$size)
  d_col <- ifelse(is.null(data$colour), "black", data$colour)
  d_fill <- ifelse(is.null(data$fill), "white", data$fill)
  d_stroke <- ifelse(is.null(data$stroke), 0.25, data$stroke)
  d_ltype <- ifelse(is.null(data$linetype), "solid", data$linetype)
  moonGrob(
    0.5, 0.5, r = d_size / 2 * 0.75, ratio = 1,
    gp = grid::gpar(
      col = scales::alpha(d_col, data$alpha),
      fill = scales::alpha(d_fill, data$alpha),
      lwd = d_stroke * ggplot2::.stroke,
      lty = d_ltype
    )
  )
}


GeomMoon <- ggplot2::ggproto(
  "GeomMoon", ggplot2::Geom,
  
  required_aes = c("x", "y"),
  
  default_aes = ggplot2::aes(
    ratio = 0.25, right = TRUE, size = 10, angle = 0,
    colour = "black", fill = "white", alpha = NA,
    stroke = 0.25, linetype = 1
  ),
  
  draw_key = draw_key_moon,
  
  draw_panel = function(data, panel_params, coord) {
    coords <- coord$transform(data, panel_params)
    moonGrob(
      coords$x, coords$y, ratio = coords$ratio, right = coords$right,
      r = coords$size / 2 * 0.75,
      angle = coords$angle,
      gp = grid::gpar(
        col = scales::alpha(coords$colour, coords$alpha),
        fill = scales::alpha(coords$fill, coords$alpha),
        lwd = coords$stroke * ggplot2::.stroke,
        lty = coords$linetype
      )
    )
  }
)


#' Moon charts
#' 
#' The moon geom is used to create moon charts, which are like pie charts except
#' that the proportions are shown as crescent or gibbous portions of a circle,
#' like the lit and unlit portions of the moon. As such, they work best with
#' only one or two groups.
#' 
#' \code{geom_moon} acts like \code{geom_point} in that multiple moons can be
#' plotted on the same panel with \code{x} and \code{y} in the plot's coordinate
#' system, but size determined independently of the coordinate system. This
#' behavior also means that the moons will always be circular even if the
#' coordinate system is not square.
#' 
#' In order to get a full circle with two complementary sections (a crescent and
#' a gibbous moon), you need to plot two shapes: one with \code{right = TRUE}
#' and one with \code{right = FALSE}, with \code{ratio} on the second one equal
#' to \code{1 - ratio} on the first.
#' 
#' @section Aesthetics:
#' \code{x} and \code{y} are required aesthetics.
#' \code{size}, \code{fill},
#' \code{colo(u)r}, \code{alpha}, \code{stroke}, and \code{group} aesthetics
#' are understood as in other \code{geom}s.
#' Two new aesthetics are also introduced: \code{ratio} and \code{right}.
#' \code{ratio} controls the proportion of the moon to be plotted, from 0 to 1.
#' \code{right} takes a boolean value to indicate whether the moon should be
#' filled from the right or the left.
#' 
#' @inheritParams ggplot2::geom_point
#' @export
#' @import ggplot2
#' @examples
#' ggplot(
#'   data.frame(x = 1:5, y = 1, size = 1:5, ratio = 1:5 * 0.2),
#'   aes(x = x, y = y, size = size, ratio = ratio)
#' ) +
#'   geom_moon()
#' 
#' # To make full moon charts, you need two calls to geom_moon(), one with
#' # right = TRUE and one with right = FALSE and ratio equal to 1 - ratio
#' # from the first one 
#' ggplot(dmeladh) +
#'   geom_moon(
#'     x = 0.5, y = 0.5, fill = "forestgreen", color = "forestgreen",
#'     aes(ratio = AdhF / 100)
#'   ) +
#'   geom_moon(
#'     x = 0.5, y = 0.5, fill = "gold", color = "gold",
#'     aes(ratio = AdhS / 100), right = FALSE
#'   ) +
#'   facet_wrap(~Locality, ncol = 7)
#'
#' # The same thing can be accomplished with a single call to geom_moon()
#' # using a "long" data frame with both frequencies if you set a grouping
#' # variable and set the `right` variable to a boolean column
#' dmeladh_long <- reshape(
#'   dmeladh,
#'   varying = c("AdhF", "AdhS"),
#'   v.names = "freq",
#'   timevar = "allele",
#'   times = c("AdhF", "AdhS"),
#'   idvar = c("Locality", "Latitude", "Longitude", "N"),
#'   direction = "long"
#' )
#' dmeladh_long$right <- rep(c(TRUE, FALSE), each = nrow(dmeladh))
#' ggplot(dmeladh_long) +
#'   geom_moon(
#'     x = 0.5, y = 0.5, key_glyph = draw_key_rect,
#'     aes(ratio = freq / 100, fill = allele, color = allele, right = right),
#'   ) +
#'   facet_wrap(~Locality, ncol = 7)
#'
#' # Moon charts (and pie charts) are sometimes useful on maps when x and y
#' # cannot be used as aesthetic dimensions because they are already spatial
#' # dimensions. Overplotting needs to be considered carefully, however.   
#' ggplot(
#'   subset(dmeladh, N > 200),
#'   aes(Longitude, Latitude)
#' ) +
#'   geom_moon(aes(ratio = AdhF / 100), fill = "black") +
#'   geom_moon(aes(ratio = AdhS / 100), right = FALSE) +
#'   coord_fixed()
geom_moon <- function(
  mapping = NULL, data = NULL, stat = "identity", position = "identity",
  na.rm = FALSE, show.legend = NA, inherit.aes = TRUE, ...
) {
  ggplot2::layer(
    geom = GeomMoon, mapping = mapping, data = data, stat = stat, 
    position = position, show.legend = show.legend,
    inherit.aes = inherit.aes, params = list(na.rm = na.rm, ...)
  )
}
